deepinterpolation/generator_collection.py

```diff
@@ -897,112 +900,144 @@ class OphysGenerator(SequentialGenerator):
+                end_ind = self.end_frame+self.post_frame+self.pre_post_omission+1
+                total_frame_per_movie = movie_obj['data'].shape[0]
+                if self.end_frame > 0 and  total_frame_per_movie > end_ind:
+                    self._movie_data = movie_obj['data'][:end_ind]
+                else:
+                    self._movie_data = movie_obj['data'][()]
+            if self.normalize_cache:
+                self._movie_data = self._movie_data.astype("float32")
+                self._movie_data = normalize(self._movie_data,
+                        self.local_mean,
+                        self.local_std)
+        return self._movie_data
+
+
+    def __gpu_cache__(self, index):
+        '''Slices minimum movie required to generate batch for a given batch
+         index and caches it onto the GPU. If a previous batch is cached, a
+         new batch is generated by only copying the difference in frames.
+        '''
+        start_ind = index*self.batch_size
+        end_ind = start_ind + self.batch_size + self.pre_frame \
+         + self.post_frame + 2*self.pre_post_omission - 2
+        if self.data_tensor is False or self.json_data["randomize"]:
+            self.data_tensor = tf.convert_to_tensor(
+                self.movie_data[start_ind:end_ind], dtype="float")
+            if not self.normalize_cache:
+                self.data_tensor = normalize(self.data_tensor, self.local_mean, self.local_std)
         else:
-            movie_obj_point = h5py.File(self.raw_data_file, "r")
-            movie_obj = movie_obj_point['data']
+            batch_frames = tf.convert_to_tensor(
+                self.movie_data[end_ind-self.batch_size:end_ind], dtype="float")
+            if not self.normalize_cache:
+                batch_frames = normalize(batch_frames, self.local_mean, self.local_std)
+            self.data_tensor = tf.concat(
+                [self.data_tensor[self.batch_size:], batch_frames], 0)
 
-        input_index = np.arange(
-            index_frame - self.pre_frame - self.pre_post_omission,
-            index_frame + self.post_frame + self.pre_post_omission + 1,
-        )
-        input_index = input_index[input_index != index_frame]
-
-        for index_padding in np.arange(self.pre_post_omission + 1):
-            input_index = input_index[input_index !=
-                                      index_frame - index_padding]
-            input_index = input_index[input_index !=
-                                      index_frame + index_padding]
-
-        # If data was cached we do not need to normalize. this was done 
-        # at once to minimize compute
-        if self.cache_data:
-            data_img_input = movie_obj[input_index, :, :]
-            data_img_output = movie_obj[index_frame, :, :]
+    def __getitem__(self, index):
+        if self.gpu_available:
+            if self.gpu_cache_full:
+                batch_indices = self.generate_batch_indexes(index)
```

Comment:
remove duplicated code block by just setting `batch_indices` and `input_indices` at top of method